Blogging Prompts: Fix Core Data concurrency issues in scheduler

dvdchr opened the pull request:
Fixes #20842

> **Warning**:
> Please note that this targets the release/22.6 branch, which is already in code freeze.

## Description

When scheduling or rescheduling prompts, the related methods make a call to `UNNotificationCenter` to either request for push authorization or fetch notification settings. However, the notification center does not guarantee that the completion block will be executed in the caller's queue. Since the completion block code involves Core Data, this raises a multithreading violation when run in debug mode.

This PR fixes the issue by wrapping the Core Data-related code within these methods in the managed object context's queue.

Additionally, this also improves the thread-safety of `BloggingPromptsService`'s init method by ensuring that any access to Core Data properties is performed within that object's context queue.

## To test

- From Xcode, launch the Jetpack app with `-com.apple.CoreData.ConcurrencyDebug 1` launch argument.
- Log in with your WordPress.com account.
- Go to Site Menu > Site Settings.
- Set Blogging Reminders with prompts enabled.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.
- Relaunch the Jetpack app from Xcode.
- 🔎 Verify that no multithreading violation is raised.
- Go to Site Menu > Site Settings.
- In the Blogging Reminders setting, uncheck all the days from the reminders.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.

Additionally, please also smoke test Blogging Reminders — creating, updating, or removing schedules; with and without prompts.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The logic is kept as close to the original as possible to minimize impact.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes, and the existing unit tests are passing.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

N/A — this PR doesn't contain UI-related changes.

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Execute Core Data code in the correct context after getting notification settings

UNNotificationCenter's `getNotificationSettings` method does not
guarantee that the completion block is called in the original queue.

This ensures that the completion block code that touches Core Data
is run within the context's queue, as obtained from the managed object.

dvdchr committed:
Execute Core Data code in the right queue after push authorization

dvdchr committed:
Improve thread safety of BloggingPromptsService init method

salimbraksa commented on WordPress/Classes/Services/BloggingPromptsService.swift:
```diff
@@ -179,17 +179,45 @@ class BloggingPromptsService {
         self.siteID = siteID
-        self.remote = remote ?? .init(wordPressComRestApi: account.wordPressComRestV2Api)
+        self.remote = remoteInstance
     }
```
@dvdchr I'm wondering if it's possible to safely access both the `siteID` and the `WPAccount` instance using just the `mainContext`. cc @crazytonyli 

```swift
var remoteInstance = remote
var siteID: NSNumber? = nil

let blogObjectID = blog?.objectID
let context = contextManager.mainContext

context.performAndWait { 
     if let blogObjectID {
       let blogInContext = try? context.existingObject(with: blogObjectID)
       siteID = blogInContext?.dotComID
     }
     let account = try? WPAccount.lookupDefaultWordPressComAccount(in: context)  
     // ...
}

// Then we set `self` properties.
```

Your approach also looks correct to me. You access the `siteID` using the blog's context and the account using the `mainContext`, that's also fine I guess. I'll let you and other reviewers decide how to move forward.

crazytonyli replied:
Agreed. We could use `coreDataStack.performQuery` function (which uses the main context) as I suggested in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854#discussion_r1228903792.

dvdchr replied:
Thanks for the input, folks 🙇🏼‍♂️. This should be addressed in https://github.com/wordpress-mobile/WordPress-iOS/commit/efdad6e4a7db74ac3503246eae8cab9f8f14f332.

crazytonyli commented on WordPress/Classes/Services/BloggingPromptsService.swift:
```diff
@@ -179,17 +179,45 @@ class BloggingPromptsService {
+            }
+            siteID = siteID ?? account.primaryBlogID
+            remoteInstance = remoteInstance ?? .init(wordPressComRestApi: account.wordPressComRestV2Api)
+        }
```
`performQuery` can return a query result. You can potentially refactor the above code to something like this (pseudo code)

```swift
let blogID = blog.objectID
let (siteID, remoteInstance) = contextManager.performQuery { context in
  let blog: Blog? = context.existingObject(with: blogID)
  let account: Account? = WPAccount.lookupDefault()
  return (
    blog.dotComID ?? account.primaryBlogID,
    .init(wordPressComRestApi: account.wordPressComRestV2Api)
  )
}
```

dvdchr replied:
Thanks. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/efdad6e4a7db74ac3503246eae8cab9f8f14f332

crazytonyli commented on WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:
```diff
@@ -43,9 +43,15 @@ class PromptRemindersScheduler {
     ///   - completion: Closure called after the process completes.
     func schedule(_ schedule: BloggingRemindersScheduler.Schedule, for blog: Blog, time: Date? = nil, completion: @escaping (Result<Void, Error>) -> Void) {
+        guard let context = blog.managedObjectContext else {
+            return
```
Do you think we should call the `completion` before returning? It's usually good to make sure completion is called once and once only.

dvdchr replied:
Good catch! This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/241e0ed5d0430a816bf76e76798cf2698b00ccff.

crazytonyli reviewed (commented):
Pattern like `blog.managedObjectContext.perform { doSomething(with: blog)}` feels like a patch to me. We probably should use it sparsely. Because there is no guarantee the `doSomething` function _only_ uses the passed-in `blog` instance in its calling thread.

For instance, a couple of call sites of `processSchedule` is changed to be called from the `blog` context thread. But the function itself accesses [the `blog` in an API callback which I believe is called from the main thread](https://github.com/wordpress-mobile/WordPress-iOS/blob/c5cf208ea8dde34b0666338efd31e12d9614d23a/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L188).

I think, yes, this change would fix the particular crash shown in https://github.com/wordpress-mobile/WordPress-iOS/issues/20848, but it may cause other issues.

One solution on top of my head is refactoring the Blog reminder classes to not use `Blog` type. It can declare a value type `struct BlogData` (with a better name probably), which includes a few property the Blog Reminder component really cares about: objectID, site id, titile, etc. So that it's not restricted by Core Data's programming model, but also has the capability to change the `Blog` model with the `objectID` if needed.

dvdchr committed:
Remove direct access to Blog after fetching in processSchedule

Refactor `addLocalNotification` and `addStaticNotifications` to
operate with value types instead of dealing with `Blog` directly.

There are still loose property reads to `Blog` within the method,
but `processSchedule` is private and its callers are guaranteeing
that the method is called from the Blog's context queue.

Further improvements will be made in the next iteration.

dvdchr committed:
Simplify the init method by utilizing the return value from performQuery

dvdchr committed:
Call the completion block when the guard condition fails

dvdchr commented:
Thanks for the input, @crazytonyli!

> [...] this change would fix the particular crash shown in https://github.com/wordpress-mobile/WordPress-iOS/issues/20848, but it may cause other issues.
> [...] One solution on top of my head is refactoring the Blog reminder classes to not use Blog type.

Great idea, and I agree. I was considering this yesterday but realized that there are going to be a lot of changes. Since I'm targeting 22.6 (which is in code freeze), I'm thinking of keeping the impact of this PR minimal and doing a larger refactor separately in the next iteration.

However, it doesn't make sense if the improvement is flawed; as you've pointed out:

> For instance, a couple of call sites of processSchedule is changed to be called from the blog context thread. But the function itself accesses [the blog in an API callback which I believe is called from the main thread](https://github.com/wordpress-mobile/WordPress-iOS/blob/c5cf208ea8dde34b0666338efd31e12d9614d23a/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L188).

Thanks for catching that. To address this, I've made some improvements to the `processSchedule` (along with a few other methods) in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/6f89aa7fd9c2d541f46cea7e486d36774bf236d7 so that the `Blog` object is no longer directly accessed from the API completion block. Although there are still some direct calls to the `Blog` object before the API call, at least the method is private and all of its callers ensure that the method is called from the `Blog`'s context queue.

In the (near) future, the larger refactor will remove the `blog` parameter from most of the Blogging Reminders and Blogging Prompts methods, replacing them with `siteID` or a value type as per your and Jeremy's suggestion.

Anyways, can you give this a second pass, please? Thank you!

crazytonyli commented on WordPress/Classes/Services/BloggingPromptsService.swift:
```diff
@@ -179,17 +179,44 @@ class BloggingPromptsService {
                    blog: Blog? = nil) {
-        guard let account = try? WPAccount.lookupDefaultWordPressComAccount(in: contextManager.mainContext),
-              let siteID = blog?.dotComID ?? account.primaryBlogID else {
+        let blogObjectID =  blog?.objectID
```
🧐 There is an extra space here.

dvdchr replied:
Good eye! 🦅. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/a9d5782042c7cc1cc58a2a755c14d2cd1b084093.

crazytonyli commented on WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:
```diff
@@ -282,12 +291,11 @@ private extension PromptRemindersScheduler {
               maxDays > 0,
-              let maxDate = Calendar.current.date(byAdding: .day, value: maxDays, to: afterDate),
-              blog.dotComID != nil else {
+              let maxDate = Calendar.current.date(byAdding: .day, value: maxDays, to: afterDate) else {
```
Not related to this PR. But I didn't realise that [`Calendar.current`](https://developer.apple.com/documentation/foundation/calendar/2293438-current)—the calendar that user picked in the device's System Settings—is used a lot in the app. In most cases, we want the Gregorian calendar, not the calendar the user likes their phone to display.

dvdchr replied:
Hmm, yeah, I think I didn't really think about it and just followed how it's done across the codebase. I'll consider fixing this during the refactor later. Thanks!

crazytonyli reviewed (approved):
Thanks for addressing my feedback. This PR looks good to me. 👍

It would be nice to see a proper refactor done to the Blog Reminder part of the app soon though 😺

dvdchr committed:
Merge remote-tracking branch 'origin/release/22.6' into fix/blogging-prompts-scheduler-concurrency

dvdchr committed:
style: remove extra spacing

mokagio commented:
@crazytonyli I ended up here while looking into something else, just wanted to say that this sounds like a great idea:

> One solution on top of my head is refactoring the Blog reminder classes to not use Blog type. It can declare a value type struct BlogData (with a better name probably), which includes a few property the Blog Reminder component really cares about: objectID, site id, titile, etc. So that it's not restricted by Core Data's programming model, but also has the capability to change the Blog model with the objectID if needed.
